Fix incorrect representation of tuples with skipped fields

Mingun opened the pull request:
This PR fixes #2105 and give some insights about the current situation with skipped fields.

You can see what changed in the generated code by running the following commands:
- place [skip.txt](https://github.com/serde-rs/serde/files/12131031/skip.txt) into `test_suite/tests` folder and change it extension to `.rs`
- run the following command:
   ```console
   ...\serde\test_suite> cargo expand --all-features --test skip > skip-N.rs
   ```
   where `N` is some number.

When commit message contains words "Changes in generated code" that means that output of the command above was changed. I recommend to diff it against the previous result and explore the difference. There will be 4 files -- baseline and three changed.

Overall changes over all commits:
- `Tuple1as0`, `Tuple1as0Default`, `Tuple1as0With`:
  - removed `visit_newtype_struct`
  - changed `*_newtype_struct` -> `*_tuple_struct(0)`
- `Tuple2as1`, `Tuple2as1Default`, `Tuple2as1With`:
  - added `visit_newtype_struct`

The table below summarizes the current behavior and the new behavior, introduced in this PR. Empty cells in the last column means that the behavior wasn't changed.

The *behavior* columns shows which methods are used to serialize or deserialize a type. The number in parenthesis is a count of fields which passed to the corresponding method.

<table>
<tr><th>Code</th><th>Current behavior</th><th>Fixed behavior</th></tr>

<tr><th colspan="3">Unit representations</th></tr>
<tr><td>

```rust
struct Unit;
```
</td><td>

`serialize_unit_struct`  
`deserialize_unit_struct`  
`visit_unit`  
</td><td></td></tr>

<tr><th colspan="3">Tuple representations</th></tr>
<tr><td>

```rust
struct Tuple0();
```
</td><td>

`serialize_tuple_struct(0)`  
`deserialize_tuple_struct(0)`  
`visit_seq`  
</td>
<td>

:information_source: It would be better to use Unit representation [(see below)](#list-of-inconsistencies)</td></tr>

<tr><td>

```rust
// Can be named
// Newtype(u8)
struct Tuple1(u8);
```
</td><td>

`serialize_newtype_struct`  
`deserialize_newtype_struct`  
`visit_newtype_struct`  
`visit_seq`  
</td><td></td></tr>

<tr><td>

```rust
struct Tuple2(u8, u8);
```
</td><td>

`serialize_tuple_struct(2)`  
`deserialize_tuple_struct(2)`  
`visit_seq`  
</td><td></td></tr>

<tr><td>

```rust
// The same as Tuple0
struct Tuple1as0(
    #[serde(skip)] u8,
);
```
#2105
</td><td>

`serialize_newtype_struct`  
`deserialize_newtype_struct`  
`visit_newtype_struct`  
`visit_seq`  
</td><td>

`serialize_tuple_struct(0)`  
`deserialize_tuple_struct(0)`  
`visit_seq`  
:information_source: It would be better to use Unit representation [(see below)](#list-of-inconsistencies)
</td></tr>

<tr><td>

```rust
// The same as Tuple0
struct Tuple2as0(
    #[serde(skip)] u8,
    #[serde(skip)] u8,
);
```
</td><td>

`serialize_tuple_struct(0)`  
`deserialize_tuple_struct(0)`  
`visit_seq`  
</td><td>

:information_source: It would be better to use Unit representation [(see below)](#list-of-inconsistencies)</td></tr>

<tr><td>

```rust
// The same as Tuple1
struct Tuple2as1(
    #[serde(skip)] u8,
    u8,
);
```
</td><td>

`serialize_tuple_struct(1)`  
`deserialize_tuple_struct(1)`  
`visit_seq`  
</td><td>

`serialize_tuple_struct(1)`  
`deserialize_tuple_struct(1)`  
`visit_newtype_struct`  
`visit_seq`  
:information_source: It would be better to use Newtype representation [(see below)](#list-of-inconsistencies)</td></tr>
</td></tr>

<tr><th colspan="3">Struct representations</th></tr>

<tr><td>

```rust
struct Struct0 {}
```
</td><td>

`serialize_struct`  
`deserialize_struct`  
`visit_seq`  
`visit_map`  
</td><td></td></tr>

<tr><td>

```rust
struct Struct1 {
    a: u8,
}
```
</td><td>

`serialize_struct`  
`deserialize_struct`  
`visit_seq`  
`visit_map`  
</td><td></td></tr>

<tr><td>

```rust
struct Struct2as0 {
    #[serde(skip)]
    a: u8,
    #[serde(skip)]
    b: u8,
}

```
</td><td>

`serialize_struct`  
`deserialize_struct`  
`visit_seq`  
`visit_map`  
</td><td></td></tr>

<tr><td>

```rust
struct Struct2as1 {
    #[serde(skip)]
    a: u8,
    b: u8,
}

```
</td><td>

`serialize_struct`  
`deserialize_struct`  
`visit_seq`  
`visit_map`  
</td><td></td></tr>
</table>


Enum variant representations
============================
The following changes also was implemented by I plan to ship them in separate PR due to some debating changes. I left my finding here to have the overall picture about skipped fields in one place.

The first debating change is to how to represent `tuple(0)` tuples? Should we represent them as a sequences with zero fields, or as a unit structs? Motivation for the unit structs is that that is the only way to skip all fields in a tuple and still represent it as a unit. This is currently done for `Enum::Tuple1as0`, but only for it. `Enum::Tuple2as0` no longer follows this agreement as well as all tuple structs, which introduces more confusion. I am inclined to this option, since to have an always empty sequence is somewhat strange.

The second debating change, maybe we should make the behavior configurable? Introduce a new attribute like `#[serde(tuple)]` to produce `tuple(0)` and `tuple(1)` forms from `TupleXas0` and `TupleXas1` tuples?

<table>
<tr><th>Code</th><th>Current behavior</th><th>Suggested behavior</th></tr>
<tr><th colspan="3">Enum::Unit representations</th></tr>
<tr><td>

```rust
enum Enum {
  Unit,
}
```
</td><td>

`serialize_unit_variant`  
`unit_variant`  
</td><td></td></tr>

<tr><th colspan="3">Enum::Tuple representations</th></tr>
<tr><td>

```rust
enum Enum {
  Tuple0(),
}
```
</td><td>

`serialize_tuple_variant(0)`  
`tuple_variant(0)`  
`visit_seq`  
</td>
<td>

`serialize_unit_variant`  
`unit_variant`  
</td></tr>

<tr><td>

```rust
// Can be called
// Enum::Newtype(u8)
enum Enum {
  Tuple1(u8),
}
```
</td><td>

`serialize_newtype_variant`  
`newtype_variant`  
</td><td></td></tr>

<tr><td>

```rust
enum Enum {
  Tuple2(u8, u8),
}
```
</td><td>

`serialize_tuple_variant(2)`  
`tuple_variant(2)`  
`visit_seq`  
</td><td></td></tr>

<tr><td>

```rust
// The same as
// Enum::Tuple0
enum Enum {
  Tuple1as0(
    #[serde(skip)] u8,
  ),
}
```
</td><td>

`serialize_unit_variant`  
`unit_variant`  
</td><td></td></tr>

<tr><td>

```rust
// The same as
// Enum::Tuple0
enum Enum {
  Tuple2as0(
    #[serde(skip)] u8,
    #[serde(skip)] u8,
  ),
}
```
</td><td>

`serialize_tuple_variant(0)`  
`tuple_variant(0)`  
`visit_seq`  
</td><td>

`serialize_unit_variant`  
`unit_variant`  
</td></tr>

<tr><td>

```rust
// The same as
// Enum::Tuple1
enum Enum {
  Tuple2as1(
    #[serde(skip)] u8,
    u8,
  ),
}
```
#2224
</td><td>

`serialize_tuple_variant(1)`  
`tuple_variant(1)`  
`visit_seq`  
</td><td>

`serialize_newtype_variant`  
`newtype_variant`  
</td></tr>

<tr><th colspan="3">Enum::Struct representations</th></tr>

<tr><td>

```rust
enum Enum {
  Struct0 {},
}
```
</td><td>

`serialize_struct_variant(0)`  
`struct_variant`  
`visit_seq`  
`visit_map`  
</td><td></td></tr>

<tr><td>

```rust
enum Enum {
  Struct1 {
    a: u8,
  },
}
```
</td><td>

`serialize_struct_variant(1)`  
`struct_variant`  
`visit_seq`  
`visit_map`  
</td><td></td></tr>

<tr><td>

```rust
enum Enum {
  // The same as
  // Enum::Struct0
  Struct2as0 {
    #[serde(skip)]
    a: u8,
    #[serde(skip)]
    b: u8,
  },
}

```
</td><td>

`serialize_struct_variant(0)`  
`struct_variant`  
`visit_seq`  
`visit_map`  
</td><td></td></tr>

<tr><td>

```rust
enum Enum {
  // The same as
  // Enum::Struct1
  Struct2as1 {
    #[serde(skip)]
    a: u8,
    b: u8,
  },
}

```
</td><td>

`serialize_struct_variant(1)`  
`struct_variant`  
`visit_seq`  
`visit_map`  
</td><td></td></tr>
</table>

<a name="list-of-inconsistencies">List of inconsistencies</a>
======================
As I said before, the current behavior contains a bunch of inconsistencies.

For example, an enum tuple variant with zero fields represented as a tuple with zero fields, but tuple variant with one skipped field represented as a unit. At the same time, tuple variant with 2 or more fields, all skipped, again represented as a tuple with zero fields. The same is applied to tuples with one effective field -- sometimes it is represented as a newtype, sometimes as a tuple with one field. I suggest to unify behavior across all cases that I summarized in the following table:

<table>
<tr><th>Value</th><th>Serialized form</th><th>Suggested form</th></tr>
<tr><th colspan="3">Types with zero fields in serialized form</th></tr>
<tr><td>

`Tuple0()`</td><td>Represented as a tuple with zero fields</td><td>unit</td></tr>
<tr><td>

`Tuple1as0(Skipped)`</td><td>:warning: Incorrectly represented as a newtype struct (bug #2105)</td><td>unit</td></tr>
<tr><td>

`Tuple2as0(Skipped, Skipped)`</td><td>Represented as a tuple with zero fields</td><td>unit</td></tr>

<tr><th colspan="3">Types with 1 field in serialized form</th></tr>
<tr><td>

`Tuple1(x)`</td><td>Represented as a newtype `x`</td><td></td></tr>
<tr><td>

`Tuple2as1(Skipped, x)`</td><td>:exclamation: Represented as a tuple with 1 field</td><td>newtype</td></tr>


<tr><th colspan="3">Enum variants with zero fields in serialized form</th></tr>
<tr><td>

`Enum::Tuple0()`</td><td>Represented as a tuple with zero fields</td><td>unit</td></tr>
<tr><td>

`Enum::Tuple1as0(Skipped)`</td><td>:exclamation: Represented as a unit struct</td><td></td></tr>
<tr><td>

`Enum::Tuple2as0(Skipped, Skipped)`</td><td>Represented as a tuple with zero fields</td><td>unit</td></tr>

<tr><th colspan="3">Enum variants with 1 field in serialized form</th></tr>
<tr><td>

`Enum::Tuple1(x)`</td><td>Represented as a newtype `x`</td><td></td></tr>
<tr><td>

`Enum::Tuple2as1(Skipped, x)`</td><td>:exclamation: Represented as a tuple with 1 field (#2224)</td><td>newtype</td></tr>
</table>

Legend:
- :warning: Bugs that leads to compilation errors
- :exclamation: Inconsistency in a block of similar table rows

Mingun committed:
Add tests for skipped fields in different kinds of structs

Enums out of scope for now, because they have too many problems

failures (1):
    tuple_struct::tuple2as1

compilation error (commented):
    tuple_struct::tuple1as0

Mingun committed:
Correctly process skipped fields in visit_newtype_struct

Changes in generated code (see the file attached to PR):
  Tuple1as0:
  Tuple1as0Default:
  Tuple1as0With:
    fixed visit_newtype_struct: use default value instead of deserializing it

This fixes compilation error for Deserialize side of Tuple1as0(Skipped) tuple

Mingun committed:
Inline `deserialize_newtype_struct`

That functions too small and used only once

(review this with whitespace changes ignored)

Mingun committed:
Move code that generates reading from a SeqAccess to a dedicated function

(review this with whitespace changes ignored)

Mingun committed:
Parameterize `read_fields_in_order` with field reading function

Mingun committed:
Change Tuple1as0(Skipped) representation: newtype -> tuple(0), similar to Tuple0() struct

A side-effect: Tuple2as1(Skipped, x) now also can be deserialized using visit_newtype_struct

Changes in generated code (see the file attached to PR):
  Tuple1as0:
  Tuple1as0Default:
  Tuple1as0With:
    removed visit_newtype_struct, *_newtype_struct -> *_tuple_struct(0)

  Tuple2as1:
  Tuple2as1Default:
  Tuple2as1With:
    added visit_newtype_struct

This commit fixes compilation error and actually fixes the issue as it was reported in #2105

dtolnay reviewed (commented):
Thanks for the PR and exemplary writeup.

Have you thought about how this direction relates to `skip_serializing_if`? For example in the following case:

```rust
#[derive(serde_derive::Serialize)]
struct Tuple(
    #[serde(skip_serializing_if = "...")]
    Option<usize>,
    #[serde(skip_serializing_if = "...")]
    Option<usize>,
    #[serde(skip_serializing_if = "...")]
    Option<usize>,
);
```

would you consider that the correct behavior needs to serialize using `serialize_newtype_struct` if exactly one of the functions returns false, and `serialize_unit_struct` if all true, and `serialize_tuple_struct` otherwise?

Mingun commented:
It is hard to say. I never used `skip_serializing_if`. At first glance I think, that we should serialize that as a tuple in all cases. That behavior will also allow to keep ability to serialize tuples with all skipped fields / only one non-skipped field as `tuple(0)` / `tuple(1)` by using `skip_serializing_if = "always_true"`, if default behavior by serializing them as `unit`s / `newtype`s would be undesirable.

In any case the behavior of `skip` should be carefully documented -- current documentation is unclear about expected results.

dtolnay commented:
I am definitely on board with the `Tuple1as0` change to fix #2105.

Not yet convinced about the `Tuple2as1` change however &mdash; I think I need to revisit this later to make up my mind about that, especially if we are going to be making `serde(skip)` intentionally inconsistent with `serde(skip_serializing_if = "always_true")`.

Mingun commented:
The problem that right now this changes are coupled and decoupling them requires to replace simple and straightforward code with something more complex. This will be especially noticeable on the part that fixes enum representations (coming soon)

dtolnay reviewed (changes requested):
I am not interested in changing the `Tuple2as1` case at this time. Individual data formats can opt to handle that by treating `serialize_tuple_struct(1)` the same as `serialize_newtype_struct`, and `deserialize_tuple_struct(1)` as `deserialize_newtype_struct`, if they prefer.

Do you see any way to separate that from the `Tuple1as0` fix?

Mingun commented:
Yes, that should be possible, just should introduce some checks and possible code duplication.

Mingun commented:
- I dropped the last commit that was changing `Tuple2as1` representation. Such structs, however, now supports deserialization using `visit_newtype_struct` method, which technically a side-effect of fixing bug and will allow to deserializers to handle  `deserialize_tuple_struct(1)` requests as `deserialize_newtype_struct`, if they prefer
- corrected tests accordingly
- add a new commit that construct all skipped fields after all stored fields. So if construction default objects is expensive they will not be constructed until needed

dtolnay commented:
I am not convinced about the last "this should increase performance" commit. Interleaving the skipped and non-skipped fields to construct them in the original source order still seems fine to me. Default objects which are expensive to construct seems like a very uncommon case to begin with, and the commit only makes any difference in the error case if I understand correctly, which is not something we micro-optimize like this.

Mingun commented:
Ok, I removed that commit. I hope compiler will able to reorder operations.

> Default objects which are expensive to construct seems like a very uncommon case to begin with

I think the code is generated also because to shift to generation taking into account all these boring uncommon cases and be sure that everything will implemented efficiently. Shifting this to the user seems to me bad practice, especially since it costs nothing.

Mingun commented:
@dtolnay , all your review comments was resolved, could you look again? I just want to remind you that all changes in the generated code are easy to see if you follow the instructions in the first message with comparing the generated code

Mingun commented:
@dtolnay, yet another gentleman's reminder...

Mingun committed:
Get rid of #deserializer_var, use more straightforward approach

Mingun committed:
Rename `deserialize_seq[_in_place]` to `read_fields_in_order[_in_place]`
